plugins/modules/system/java_keystore.py

```diff
@@ -388,6 +422,7 @@ def cryptography_create_pkcs12_bundle(self, keystore_p12_path, key_format='PEM',
 
         with open(keystore_p12_path, 'wb') as p12_file:
             p12_file.write(pkcs12_bundle)
+        return dict(msg='', cmd='python cryptography serialize_key_and_certificates()', rc=0)
```

Comment:
If the `result` dict is passed around methods it would make sense to add a `self._result` attribute to the class.

Reply:
In fact, I've added a line to track a change: the removal of the existing keystore (in `JavaKeystore.create()`), and so I needed a `result` to return, with the `changed` key in.

But I'm not sure this is the right way, because:
- this adds a `changed` key in failed results, that is not very common, and
- the module can fail *after* removing the keystore, that is not very wishable (probably the reason why it is *not very common*)

Reply:
I would agree that decoupling the kwargs passed to `exit_json` and the kwargs passed to `fail_json` is a more flexible approach. It was more coincidence that there was overlap in this case.

Reply:
What is not satisfying for me is that we remove the existing keystore, and then try to create a new one. If the creation of the new keystore fails (and is is expected that it may fail, since there is at least one `module.fail_json()` here), the user ends up with no keystore at all, whereas he started with a (maybe valid) keystore. I think this is not the expected behavior for an ansible module.

I think we shouldn't remove the existing keystore before creating the new one. Fixing this is not related to this PR; on the other hand, the simple add of a `changed=True` at keystore removal, finally bloats the module anyway, just to track this change. What I mean is that I should remove all this `result` stuff and keep all in place (early removal of the keystore without notifying about that), or remove all this `result` stuff and fix this *too-early-removal* issue in *this* PR.

@Ajpantuso @felixfontein @russoz any thought about this ?

Reply:
> What is not satisfying for me is that we remove the existing keystore, and then try to create a new one. If the creation of the new keystore fails (and is is expected that it may fail, since there is at least one `module.fail_json()` here),

Every external invocation can fail (for very different reasons). Not having a `fail_json` (or indirect `check_rc=True`) would be a bug, even if we don't expect a failure.

> the user ends up with no keystore at all, whereas he started with a (maybe valid) keystore. I think this is not the expected behavior for an ansible module.

That's true. Depending on the module this is impossible to avoid anyway. In this case, we can try to minimize the impact: create the new keystore in a temp location, and then use `module.atomic_move()` to move it to the final location. Obviously the atomic move can fail, or can turn out to not be atomic, but the chance that something happens exactly in that code is a lot less.

> I think we shouldn't remove the existing keystore before creating the new one. Fixing this is not related to this PR; on the other hand, the simple add of a `changed=True` at keystore removal, finally bloats the module anyway, just to track this change. What I mean is that I should remove all this `result` stuff and keep all in place (early removal of the keystore without notifying about that), or remove all this `result` stuff and fix this _too-early-removal_ issue in _this_ PR.

What's often done: keep a `changed` variable in the class which is set to `True` when a change happens (`self.changed = True`) and which in the end is put into result. Also you could have `self.result = {}` which is updated when needed. (You could also use `self.result['changed']` instead of a separate `self.changed`.)

Reply:
I am very off this context, but to me it sounds like a `backup` option in the module, with a default `True`, for keeping a copy of the old file. Once that is done, remove away and don look back.